openzfs 6739 - assumption in cv_timedwait_hires

rtvd opened the pull request:
Porting from openzfs

Userland version of cv_timedwait_hires() always assumes absolute time.
Reviewed by: Paul Dagnelie <pcd@delphix.com>
Reviewed by: Matthew Ahrens <mahrens@delphix.com>
Reviewed by: Dan McDonald <danmcd@omniti.com>
Reviewed by: Robert Mustacchi <rm@joyent.com>
Approved by: Robert Mustacchi <rm@joyent.com>

behlendorf commented:
This patch results in ztest failures on Linux.  We'll need to investigate why this happens in Linux and presumably not under OpenZFS.

```
ASSERT at kernel.c:563:cv_timedwait_hires()zdb: can't open 'ztest': No such file or directory
```

rtvd commented:
I suppose the relevant ASSERT statement is this one:

ASSERT(error == 0);

Is there, perhaps, an alternative form of ASSERT which would print the actual value of error?

rtvd commented:
That "error" is returned by a call to pthread_cond_timedwait.
The two possible results are checked for: ETIME and EINTR.
However, according to http://linux.die.net/man/3/pthread_cond_timedwait EINTR cannot be returned at all:
"These functions shall not return an error code of [EINTR]."
Also, ETIME is not mentioned.
Instead, the possible error codes are ETIMEDOUT, EINVAL and EPERM.

behlendorf commented:
> Is there, perhaps, an alternative form of ASSERT which would print the actual value of error?

In new code we encourage the use of ASSERT0() which will log the actual value if it's not 0.  There's also ASSERT3S(), ASSERT3U(), and ASSERT3P() which have the same behavior.  I'd suggest updating that last assert to `ASSERT0(error)`.

I took a quick look at this and it looks like you've just stumbled across something really interesting.  So the illumos/freebsd code was built on the `cond_*` interface for their platform.  For Linux that didn't make any sense so instead we build these functions on top of pthread condition variables.  And while the semantics are almost identical the return values are not.  Here are the respective man pages:

Illumos: [cond_reltimedwait()](https://docs.oracle.com/cd/E53394_01/html/E54766/cond-reltimedwait-3c.html) man page
Linux: [pthread_cond_timedwait()](http://linux.die.net/man/3/pthread_cond_timedwait) man page

What you're going to need to do is update the error code handling in this function to match the expected `pthread_cond_timedwait()` return values.  I'm a little surprised this hasn't caused more trouble in ztest up to this point.

rtvd commented:
I have added a commit which should fix that code. However, I am not sure where should that patch have gone and how it should have been called.
Presently it is the second commit in openzfs-6739 branch in my fork of zfs.

https://github.com/rtvd/zfs/commit/e20f2b9450ee4466040bb38b1652ffd111f86f72

rtvd commented:
I think reviewing the rest of pthread_* makes sense. I can do that.
If I do it, should I raise an issue for that and do it in a separate branch?

behlendorf commented on lib/libzpool/kernel.c:
```diff
@@ -481,9 +481,7 @@ cv_wait(kcondvar_t *cv, kmutex_t *mp)
-	int ret = pthread_cond_wait(&cv->cv, &mp->m_lock);
-	if (ret != 0)
-		VERIFY3S(ret, ==, EINTR);
+	VERIFY3S(pthread_cond_wait(&cv->cv, &mp->m_lock), ==, 0);
```
`VERIFY0(error)` is the same as `VERIFY3S(error, ==, 0)` and preferred for this kind of usage.

behlendorf commented:
Minor build issue with the updated patch:

```
kernel.c:536:1: error: label ‘top’ defined but not used [-Werror=unused-label]
```

The changes themselves look good to me I'd just recommend switching to VERIFY0.

> I think reviewing the rest of pthread_* makes sense. I can do that.

Thanks, I think that's a really good idea.  We could easily have made similar mistakes elsewhere.

rtvd commented:
OK. I will switch it to VERIFY0 and review the rest.

rtvd commented:
Pushed one more commit.

rtvd commented:
@behlendorf , I have updated my branch (rebased on master and squashed the commits). Also the "top:" label is gone.
However, it cannot be cleanly compiled because of this:

> arc.c:3509:4: error: implicit declaration of function ‘cv_timedwait_sig_hires’ [-Werror=implicit-function-declaration]

Which is also affecting the master.

behlendorf commented:
You need to update to the latest version of the spl to resolve the build issue.

rtvd commented:
Thank you. Updating spl did resolve the problem.
Good enough for a merge?

behlendorf commented:
@rtvd yup, LGTM!  Thanks for refreshing this and doing the pthreads cleanup.
